upb/upb.c

```diff
@@ -239,11 +255,13 @@ bool upb_arena_addcleanup(upb_arena *a, void *ud, upb_cleanup_func *func) {
   return true;
 }
 
-void upb_arena_fuse(upb_arena *a1, upb_arena *a2) {
+bool upb_arena_fuse(upb_arena *a1, upb_arena *a2) {
   upb_arena *r1 = arena_findroot(a1);
   upb_arena *r2 = arena_findroot(a2);
 
-  if (r1 == r2) return;  /* Already fused. */
+  if (r1 == r2) return true;  /* Already fused. */
```

Comment:
Maybe it would be simplest and safest right now if we also make the fuse fail if either arena has a value other than `&upb_alloc_global` for the allocator.

Any custom allocator could have the same problem as the initial block: it could have a hard lifetime bound that cannot be extended arbitrarily.

Reply:
I thought about this, but anything that presents an allocator interface really is signing up to let things live until the free function is called.

Reply:
The allocation function takes an arbitrary pointer to user data, I think the user is only signing up to make that live until `upb_arena_free()`.

On the other hand, I am not sure anybody is using the custom allocator right now. So this is all a bit theoretical until somebody is.

One thing that is clear is that the current logic is busted if you fuse arenas that use different allocators. So rejecting anything but the global allocator is an easy short term fix that will at least prevent broken logic.

Reply:
Added a restriction so that fuse only works on block allocs